x-pack/plugins/siem/public/components/url_state/index.tsx

```diff
@@ -0,0 +1,439 @@
+  }: UrlStateContainerLifecycleProps) {
+    const { history, location, urlState } = this.props;
+    if (JSON.stringify(urlState) !== JSON.stringify(prevUrlState)) {
+      Object.keys(urlState).forEach((key: string) => {
+        if ('kqlQuery' === key || 'timerange' === key) {
+          const urlKey: KeyUrlState = key;
+          if (JSON.stringify(urlState[urlKey]) !== JSON.stringify(prevUrlState[urlKey])) {
+            if (urlKey === 'kqlQuery') {
+              urlState.kqlQuery.forEach((value: KqlQuery, index: number) => {
+                if (
+                  JSON.stringify(urlState.kqlQuery[index]) !==
+                  JSON.stringify(prevUrlState.kqlQuery[index])
+                ) {
+                  this.replaceStateInLocation(urlState.kqlQuery[index], 'kqlQuery');
+                }
+              });
+            } else {
+              this.replaceStateInLocation(urlState[urlKey], urlKey);
+            }
+          }
+        }
+      });
+    }
+
+    if (history.action === 'POP' && location !== prevLocation) {
+      this.handleLocationChange(prevLocation, location);
+    }
+  }
+
+  public componentDidMount() {
+    const { location } = this.props;
+    this.handleInitialize(location);
+  }
+
+  private replaceStateInLocation = throttle(
+    1000,
+    (urlState: UrlInputsModel | KqlQuery | KqlQuery[], urlStateKey: string) => {
+      const { history, location } = this.props;
+      const newLocation = replaceQueryStringInLocation(
+        location,
+        replaceStateKeyInQueryString(urlStateKey, urlState)(getQueryStringFromLocation(location))
+      );
+
+      if (newLocation !== location) {
+        history.replace(newLocation);
+      }
+    }
+  );
+
+  private urlStateMappedToActions: UrlStateMappedToActionsType = {
+    kqlQuery: {
+      hosts: this.props.setHostsKql,
+      network: this.props.setNetworkKql,
+    },
+    timerange: {
+      absolute: this.props.setAbsoluteTimerange,
+      relative: this.props.setRelativeTimerange,
+    },
+  };
+
+  private handleInitialize = (location: Location) => {
+    Object.keys(this.urlStateMappedToActions).forEach((key: string) => {
+      if ('kqlQuery' === key || 'timerange' === key) {
+        const urlKey: KeyUrlState = key;
+        const newUrlStateString = getParamFromQueryString(
+          getQueryStringFromLocation(location),
+          urlKey
+        );
+        if (newUrlStateString) {
+          if (urlKey === 'timerange') {
+            const timerangeStateData: UrlInputsModel = decodeRisonUrlState(newUrlStateString);
+            const globalId: InputsModelId = 'global';
+            const globalRange: UrlTimeRange = timerangeStateData.global;
+            const globalType: TimeRangeKinds = get('global.kind', timerangeStateData);
+            if (globalType) {
+              if (globalRange.linkTo.length === 0) {
+                this.props.toggleTimelineLinkTo({ linkToId: 'global' });
+              }
+              // @ts-ignore
```

Comment:
There are 2 remaining ts issues, this is one of them. I've spent the better part of the past 6 hours trying to figure this out. May need to merge and come back. Even after adding `UrlStateMappedToActionsType` and wrapping it in another if, it still gives this error:
<img width="776" alt="Screen Shot 2019-05-21 at 9 42 41 AM" src="https://user-images.githubusercontent.com/6935300/58121136-e9a11d00-7bc3-11e9-88df-8c1f8051c9a7.png">